### ISSUE TITLE
Single Notification Handler

### ISSUE THREAD

**Issue by gotev** [gotev/android-upload-service#494]
## If you're reading this when this issue is closed, update to 4.3.0+ and visit [Notification Handler Wiki](https://github.com/gotev/android-upload-service/wiki/Configuration#notification-handler)

This is a `hot` topic. How to show a single notification for all the uploads? 

After many requests 🔥and no contributions so far 😿, this issue aims at finding a permanent solution to be included in the next Upload Service release.

> This is only for latest 4.x Upload Service release. No backports.

> DISCLAIMER: The following code is feedback-ware. If you use it, leave a feedback in this issue. I've done it rapidly and tested it in some simple use-cases with up to 4 parallel uploads. It's pretty rough at this time. No rate limiting or other advanced features. It can be improved. Suggestions are welcome.

Add the following file to your project:

`AbstractSingleNotificationHandler.kt`
```kotlin
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import net.gotev.uploadservice.UploadService
import net.gotev.uploadservice.UploadServiceConfig
import net.gotev.uploadservice.data.UploadInfo
import net.gotev.uploadservice.data.UploadNotificationConfig
import net.gotev.uploadservice.data.UploadNotificationStatusConfig
import net.gotev.uploadservice.exceptions.UserCancelledUploadException
import net.gotev.uploadservice.network.ServerResponse
import net.gotev.uploadservice.observer.task.UploadTaskObserver
import java.util.concurrent.ConcurrentHashMap

abstract class AbstractSingleNotificationHandler(
    private val service: UploadService,
    internal val notificationChannelId: String
) : UploadTaskObserver {

    enum class TaskStatus {
        InProgress,
        Succeeded,
        Failed,
        Cancelled
    }

    data class TaskData(
        val status: TaskStatus,
        val info: UploadInfo,
        val config: UploadNotificationStatusConfig
    )

    private val tasks = ConcurrentHashMap<String, TaskData>()

    private val notificationManager by lazy {
        service.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    }

    init {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager.getNotificationChannel(notificationChannelId)
                ?: throw IllegalArgumentException("The provided notification channel ID $notificationChannelId does not exist! You must create it at app startup and before Upload Service!")
        }
    }

    fun removeTask(uploadId: String) {
        tasks.remove(uploadId)
    }

    abstract fun updateNotification(
        notificationManager: NotificationManager,
        notificationBuilder: NotificationCompat.Builder,
        tasks: Map<String, TaskData>
    ): NotificationCompat.Builder?

    @Synchronized
    private fun updateTask(
        status: TaskStatus,
        info: UploadInfo,
        config: UploadNotificationStatusConfig
    ) {
        tasks[info.uploadId] = TaskData(status, info, config)
        val builder = NotificationCompat.Builder(service, notificationChannelId)
        val notification = updateNotification(notificationManager, builder, HashMap(tasks))
            ?.setGroup(UploadServiceConfig.namespace)
            ?.setOngoing(true)
            ?.build()
            ?: return
        service.holdForegroundNotification(javaClass.name, notification)
    }

    override fun onStart(
        info: UploadInfo,
        notificationId: Int,
        notificationConfig: UploadNotificationConfig
    ) {
        updateTask(TaskStatus.InProgress, info, notificationConfig.progress)
    }

    override fun onProgress(
        info: UploadInfo,
        notificationId: Int,
        notificationConfig: UploadNotificationConfig
    ) {
        updateTask(TaskStatus.InProgress, info, notificationConfig.progress)
    }

    override fun onSuccess(
        info: UploadInfo,
        notificationId: Int,
        notificationConfig: UploadNotificationConfig,
        response: ServerResponse
    ) {
        updateTask(TaskStatus.Succeeded, info, notificationConfig.success)
    }

    override fun onError(
        info: UploadInfo,
        notificationId: Int,
        notificationConfig: UploadNotificationConfig,
        exception: Throwable
    ) {
        if (exception is UserCancelledUploadException) {
            updateTask(TaskStatus.Cancelled, info, notificationConfig.cancelled)
        } else {
            updateTask(TaskStatus.Failed, info, notificationConfig.cancelled)
        }
    }

    override fun onCompleted(
        info: UploadInfo,
        notificationId: Int,
        notificationConfig: UploadNotificationConfig
    ) {
    }
}
```

Then implement your custom single notification handler. Example:

`MyNotificationHandler.kt`
```kotlin
class MyNotificationHandler(service: UploadService, notificationChannelId: String)
    : AbstractSingleNotificationHandler(service, notificationChannelId) {
    override fun updateNotification(
        notificationManager: NotificationManager,
        notificationBuilder: NotificationCompat.Builder,
        tasks: Map<String, TaskData>
    ): NotificationCompat.Builder? {
        // Return null to not update the notification
        return notificationBuilder
            .setContentTitle("${tasks.size} Uploads")
            .setContentText("${tasks.values.count { it.status == TaskStatus.InProgress }} in progress")
            .setSmallIcon(android.R.drawable.ic_menu_upload)
    }
}
```

To set it up, add this in your `Application`:

```kotlin
UploadServiceConfig.notificationHandlerFactory = { service ->
    MyNotificationHandler(service, "yourChannel")
}
```
Remember to call `removeTask(uploadId)` when you don't need to show a certain uploadTask information. The map contains all the uploads (current and done) performed, to allow diverse UI displays, but it has the downside to grow indefinitely. For this reason, the method `removeTask` has been added.

**Comment by gotev** [comment]
@nickadna you can try this one

**Comment by nickadna** [comment]
@gotev cool 👍

**Comment by gotev** [comment]
@nickadna Remember to call removeTask(uploadId) when you don't need to show a certain uploadTask information. The map contains all the uploads (current and done) performed, to allow diverse UI displays, but it has the downside to grow indefinitely. For this, the method removeTask has been added. Not sure it's the best way to deal with this.

Updated the first post as well.

I'm open to suggestions about that.

**Comment by nickadna** [comment]
Yes, but as long as we will remove completed tasks it will pass. We had nothing to worry about.

**Comment by gotev** [comment]
@nickadna updated `AbstractSingleNotificationHandler`. Now it passes only the proper `UploadNotificationStatusConfig` of each task instead of all the configuration and returns a copy of the original task map to prevent `ConcurrentModificationException` in case you are calling `removeTask` in a loop.

**Comment by nickadna** [comment]
@gotev great

**Comment by gotev** [comment]
This is going to be embedded in the next release, estimated before the end of this week.

**Comment by gotev** [comment]
Available in `4.3.0`. Visit [Notification Handler Wiki](https://github.com/gotev/android-upload-service/wiki/Configuration#notification-handler)
